### ISSUE TITLE
Consider adding a few more standard comparison guards

### ISSUE THREAD

**Issue by girtsl** [safakgur/guard#40]
**Is your feature request related to a problem?**
Some standard validations present in other validation libraries seem to missing.

**Describe the solution you'd like**
New comparison guards:
- `GreaterThan(T)`
- `GreaterThanOrEqualTo(T)` - an alias to `Min(T)` or rather the other way around
- `LessThan(T)`
- `LessThanOrEqualTo(T)` - an alias to `Max(T)` or rather the other way around

New `DateTime` guards:
- `Before(T)`
- `NotBefore(T)` or `NoSoonerThan(T)`
- `After(T)`
- `NotAfter(T)` or `NoLaterThan(T)`

Note that the default [DateTime.Compare()](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.compare?view=netcore-3.0) method ignores the `Kind` property and the comparison itself is useless if the values are not in the same time zone. I don't expect these new proposed `DateTime` guards to take into account every aspect of datetime comparison (https://blog.nodatime.org/2011/08/what-wrong-with-datetime-anyway.html) but they could at least make sure the `Kind` properties match on both values before comparison and fail if that's not the case. My hope is that this would force the developers to make a conscious choice and force all `DateTime` values to have the same `Kind`. Different time zones are not considered because anyone working with explicit time zone support likely doesn't use `DateTime` anyway.

I could also create a PR if you consider these guards worth adding.

**Comment by safakgur** [comment]
Great suggestions, thank you.

The reason I only included the inclusive checks first was that in a lot of codebases the limits are almost always stored inclusively, like `MaxValue`, `MinAge`, `MaxNameLength`, etc. And the reason I picked the current names was that I found `.Max(5)` and `.MinLength(3)` to be more readable than `.GreaterThanOrEqualTo(5)` and `.LengthGreaterThanOrEqualTo(3)`. I believe they especially make more sense when used like `.Min(MinValue)`. I never needed exclusive checks myself when validating an argument, but they do seem common enough that I understand why you may need those. We can consider adding `LessThan` and `GreaterThan`.

I am strongly against having two guards that do exactly the same thing, so inclusive checks should either stay as they are or get renamed to their longer forms (would require a major version to remove `Min` and `Max` but we can just deprecate them first in a minor one). But my preference would be to keep them as they are as they cover the majority of use cases and are imho much more readable this way. I'd even go for `MinExclusive` and `MaxExclusive` for the new guards if we really had to name them alike.

I agree that we should have guards to compare dates but I rather have Min/Max/LessThan/GreaterThan overloads that accept date arguments than to have new, date-specific names - this would also prevent users from accidentally calling the wrong API. Comparing local and UTC dates should be alright as the framework does the necessary conversion. For example, Json.NET would deserialize "2019-12-10T22:25:20Z" as UTC time and the user can compare it with both `DateTime.Now` and `DateTime.UtcNow` to the same effect. I believe the real problem here is `DateTimeKind.Unspecified` as the framework happily treats it like local time. 

What do you think?

**Comment by girtsl** [comment]
I guess this is one of those things where it all depends on the context or personal preference. No matter which option you pick I can always define my own aliases that I find more appropriate. The following is my opinion so it's perfectly fine to disagree.

I'm sure there are many uses for exclusive bounds but for me they come into picture when dealing with fractions. For instance, I don't care so much about the actual value as long as it's above a certain threshold.

`Min()` and `Max()` usually represent **functions** to **find** the extremes of a given set whereas `GreaterThan(value)` and `LessThan(value)` - comparisons or assertions, which is what we actually want. While `Min(value)` makes sense in the context of this particular library, it's quite confusing otherwise, especially if the codebase itself uses the functions `Min()` and `Max()`.

I also believe it's important to be consistent and provide an intuitive API. Having `Min()`, `Max()`, `LessThan()` and `GreaterThan()` will be more confusing than any of the other two options: 
- Min, Max, MinExclusive, MaxExclusive
- GreaterThan, LessThan, GreaterThanOrEqualTo, LessThanOrEqualTo.

Personally, I think `MinExclusive` and `MaxExclusive` are eyesores and there is no immediate association to what are they supposed to do. They don't read as a sentence, too: "age minimum exclusive is 18". 
Two other great libraries ([FluentValidation](https://fluentvalidation.net/built-in-validators) and [FluentAssertions](https://fluentassertions.com/numerictypes/)) that I use for input validation and unit tests respectively use style similar to what I have suggested above. In fact, I might have been conditioned by them as I've used them for several years. Of course, that doesn't mean you should follow the suit.

When it comes to specializations (or shortcuts/aliases in some cases), I think they are perfectly fine and provide better semantics. Even this library has them:
- `Positive()` == `Min(0)`
- `NotPositive` == `Max(0)`
- `MinLength()` == `Member(_ => _.Length.Min())`
- etc.

That being said, a validation like `Positive()` is perfectly suitable for *numbers*. For other values that carry a certain semantic meaning but are expressed as numbers, e.g. count, length, amount, that validation doesn't really "fit":
```csharp
Guard.Argument(() => count).Positive(); 
Guard.Argument(() => count).Min(0);
Guard.Argument(() => discount).Positive(); // a discount, if present, will always be "positive" - that's implied - but we need to assert it's greater than 0
```
vs
```csharp
Guard.Argument(() => count).GreaterThan(0); 
Guard.Argument(() => discount).GreaterThan(0); // for the sake of simplicity discount is assumed to be a number whereas normally I would use a `Money`, `Percentage` or even `Discount` value object
```

It's similar with dates, it just doesn't feel natural to do this:
```csharp
Guard.Argument(() => timestamp).Min(DateTime.UtcNow);
Guard.Argument(() => timestamp).GreaterThan(DateTime.UtcNow);
```
if I could do this instead:
```csharp
Guard.Argument(() => timestamp).OnOrAfter(DateTime.UtcNow);
Guard.Argument(() => timestamp).NotBefore(DateTime.UtcNow);
```

Now, the date validation concerns, which, looking back, should have been put in their own issue.
The default `Kind` in .NET is `Unspecified` which json.net will *serialize* without the time zone designator. If you store a `DateTime` value (even with the `Kind` explicitly set to, say, `Utc`) in MS SQL Server and use the `datetime` data type instead of `datetimeoffset` it will be read back as `Unspecified`. These gotchas, which are very common, can cause all kinds of nasty bugs if you are not aware of them.

**Comment by safakgur** [comment]
The following examples are not direct aliases as they accept different sets of parameters:

> `Positive()` == `Min(0)`
> `NotPositive` == `Max(0)`
> `MinLength()` == `Member(_ => _.Length.Min())`

Even the ones that look equal like `NotDefault` and `NotZero` are actually different as `NotDefault` uses `EqualityComparer<T>.Default.Equals` and is available for every struct (e.g. Guid) while `NotZero` uses `IComparable<T>.CompareTo` and is available only for the structs that implement `IComparable<T>` (numerics).  So even if it's subtle, there is always a difference in either the way you call the validation or the implementation it uses.

The lack of precedence doesn't mean we should never do this - I hear your reasoning about consistency but I'm not yet convinced that it's worth adding eight overloads that will add no new functionality:

* `Min`: `GreaterThanOrEqualTo` - 2
* `MinLength`: `LengthGreaterThanOrEqualTo` - 1
* `MinCount`: `CountGreaterThanOrEqualTo` - 1
* `Max`: `LessThanOrEqualTo` - 2
* `MaxLength`: `LengthLessThanOrEqualTo` - 1
* `MaxCount`: `CountLessThanOrEqualTo` - 1

I believe "whichever _looks_ more appropriate" is not a good enough answer to the question "which one should I use?". People will remember that the equivalent of `<= 5` is `Max(5)` after they use it once but they will never truly grasp the difference -or the lack of it- between `Max(5)` and `LessThanOrEqualTo(5)`, and when to prefer which.

I believe we should add `GreaterThan` and `LessThan` guards but after that we have only two viable options:

* Leave `Min`/`Max` overloads as they are.
* Rename them to `GreaterThanOrEqualTo`/`LessThanOrEqualTo`

The latter will require a major version bump anyway, so let's add the new ones first and we can discuss renaming the existing ones in a separate issue.

---

The problem with using DateTime-specific names is the fact that it would still be possible to do `Guard.Argument(() => ts).Min(DateTime.UtcNow)` as the existing comparison guards work with every type that implements `IComparable<T>`. And some people will discover `Min`/`GreaterThanOrEqualTo` before `OnOrAfter` when they're going through intellisense.

So we must have a `Min`/`GreaterThanOrEqualTo` overload that accepts a `DateTime` instead of `T` at the very least for these people to fall into the pit of success. And if we're already doing that, adding an `OnOrAfter` alias because it conveys the meaning just a little bit better doesn't make much sense to me as I think it would confuse more people than it helps.

I propose us to:

* Add a new guard, `KindSpecified`, to require a `DateTime` argument to have either `DateTimeKind.Local` or `DateTimeKind.Utc`, so people can force callers to think about the source of their time values and specify the kinds explicitly.
* Add non-generic overloads to `Equal`, `NotEqual`, `Min`, `Max` and `InRange` that accept `DateTime` arguments and throw unless `a.Kind == b.Kind || (a.Kind == Unspecified) == (b.Kind == Unspecified)`, allowing comparisons between local and UTC times.

**Comment by girtsl** [comment]
I never suggested to get rid of `MinLength`, `MinCount`, and their `Max*` counterparts - I think they are perfectly fine and represent very common concepts. It's only the `Min` and `Max` methods I would want to see replaced by `GreaterThanOrEqualTo` and `LessThanOrEqualTo` respectively.

---

When it comes to the shortcuts/aliases I brought up, I'll admit, I only looked at the API surface without paying too much attention on what kind of arguments those methods accept. Having now looked more closely at the source code, I think there are a couple more semantic/design issues.

I think you'll agree that for numbers the methods `Positive()` and `Min(0)` behave in the same way - and this is to be expected otherwise there would be a big problem. I wrote my previous comments with specific use cases in mind - when `T` is a number - and it didn't occur to me that a method like `Positive()` could be applied to a non-number - which is an issue by itself.

IMO some of the methods make claims that don't make sense if the `T` is not a number:
```csharp
/// <summary>Requires the argument value to be zero.</summary>
ArgumentInfo<T> Zero<T>(...) where T : struct, IComparable<T>
{
    if (Comparer<T>.Default.Compare(argument.Value, default) != 0)
    //...
}

/// <summary>Requires the argument to have a value that is greater than zero.</summary>
ArgumentInfo<T> Positive<T>(...) where T : struct, IComparable<T>
{
    if (Comparer<T>.Default.Compare(argument.Value, default) <= 0)
    //...
}

//etc.
```
Those claims are false because the methods don't compare the **argument** to the number `0` but the result of invoking the default comparer.
Consider a struct like this:
```csharp
public struct Foo : IComparable<Foo>
{
	public string Bar;
	public int CompareTo(Foo other)
	{
		...
	}
}

// and the argument checks:
Guard.Argument(() => foo).Positive(); // what does this even mean?
Guard.Argument(() => foo).NotZero(); // huh? This would imply that there is an implicit conversion from the struct to 0, but there isn't.
```

My personal opinion is that your aim here has been a tad too wide. Methods like `Zero`, `NotZero`, `Positive`, etc. should only be defined for *numbers* and not for every struct that implements `IComparable<T>`. OTOH, methods like `GreaterThan`, `LessThan`, etc. make perfect sense because it's clear we are comparing the argument to a value of the same type and there are well defined comparison rules.

---

I guess methods for DateTime guards with better semantics are not easily achievable, at least not without being very specific about the specification of `T` for each method (as FluentValidation have done it). 
Your proposals are sound and do appear to solve the issues I outlined in the original comment.

---

I hope you don't take my criticism the wrong way. I think this is a great library and it has helped me make the code a bit tidier and more expressive. My intent is to help make it even better or at least provide an alternative point of view by sharing my experience.  Thank you for your great work!

**Comment by safakgur** [comment]
I apologize for the very late response. I also want to say that I really appreciate your contributions whether in the form of ideas, criticisms or pull requests. I'll be creating separate issues to track GreaterThan/LessThan and kind-aware DateTime guards so we can start implementing.

About comparable struct guards, there are a couple things that led me to design them as they are:

* There is no way to constrain a generic parameter to be a "number type" as the number types in .NET do not have a shared interface like `INumeric`.
* Structs _should_ have valid and useful zero values as they can't have default constructors. Even as this is not always the case (e.g. Guid), the best practice is to choose structs where a zero-inited value makes sense (e.g. numbers, flag enums)
* I wanted to allow custom comparables that adhere to the above best practice (e.g. a Money struct) to use guards like Zero, Positive and Negative.

I agree that there are cases where this doesn't make sense, but I think the current solution covers the 99% as A - Structs that implement `IComparable<T>` are most likely to be number-like types (in fact I can't think of an `IComparable<T>` struct that is not a number) and B - We can actually trust the user not to call `colorArg.Negative()`. The trade-off here is a little intellisense pollution which I had found acceptable at the time.

If you think of a use case where these guards would more likely to confuse people than help them, we can consider constraining the type further with `IComparable, IComparable<T>, IConvertible, IEquatable<T>, IFormattable` in order to reduce its possibility of showing up for types except built-in numbers.

**Comment by girtsl** [comment]
It's now my turn to apologize for not replying in a timely manner - work and, well, life have been my top priorities.

- Yes, there is no "numeric" type constraint for generic parameters hence methods like `Zero`, `NotZero`, etc. are quite ambiguous which IMO makes them a design smell. It's only my opinion of course. 
Apparently there is a good reason for that - https://stackoverflow.com/questions/32664/is-there-a-constraint-that-restricts-my-generic-method-to-numeric-types.
- Can you please point me to a resource where such best practice is described? Do you perhaps mean a default value? To me they mean two very different things. 

I can think of a number of examples where the type could implement `IComparable` but it would not be "convertible" to 0: 
- `Money` - 0 amount but in what currency?
- `CountryCode` - if you use one of the ISO 3166-1 standards, countries have numeric codes but none have a code 0.
- `Language` - ISO 639-* don't define numeric codes.
- Any other value object that could be shown in an ordered collection.  

Yes, it's quite natural that a "numeric" struct would also implement `IComparable` but it's more of a side-effect of the interface's name and the `CompareTo` method's semantics.

I guess the main issue is that the name `IComparable` is a bit ambiguous but the [interface ](https://docs.microsoft.com/en-us/dotnet/api/system.icomparable-1?view=netcore-3.1) itself is only meant to define a method to be used when *ordering* or *sorting* instances of the same type, it does not make sense to have a general comparison to the number 0. Of course, it makes sense for structs that can be "converted"/"folded" to a number but I'd say it's more of an exception than a rule. I hope you now see why methods like `GreaterThan`, `LessThan`, etc. make sense but methods like `Zero`, `Positive`, etc. don't. 

You will most often see `IComparable` be used on types that have *string* representation (and thus are shown somewhere, perhaps in an ordered way), `Guid` (UUID) is a perfect example of that. Even though the UUID is defined as a 128-bit number and has a special "nil" case, it doesn't make sense to treat it as number. 
With all this in mind, I think there is a value in defining constructors such as `ArgumentInfo<int>`, `ArgumentInfo<long>`, etc. that could have methods such as `Zero`, `Positive`, etc.

**Comment by safakgur** [comment]
Welcome back :)

I agree with all your points, but I also believe what we have in place is a reasonable compromise.

What we have:

* ✔️ Less, easier to manage code
* ✔️ Support for client-defined types
* ❌ Intelli-sense pollution for non-numeric types

Having byte/sbyte/short/ushort/int/uint/long/ulong/float/double/TimeSpan overloads instead:

* ❌ Probably a T4 template to implement without going mad
* ❌ No support for client-defined types
* ✔️ It's not possible for someone to do `Argument(() => dateOfBirth).Positive()`

**The longer part:**

> Can you please point me to a resource where such best practice is described?

The [structs page][1] in the spec has "Structs should be designed to consider the default initialization state a valid state" and the [design guidelines][2] has "DO ensure that a state where all instance data is set to zero, false, or null (as appropriate) is valid."

> Do you perhaps mean a default value? To me they mean two very different things.

Why is that? The default value of a struct have all of its fields zero-inited, i.e. all of its bytes will be zero (which gives you null for references).

> * `Money` - 0 amount but in what currency?

Mine uses XXX\999 (no currency) as default. Comparison and arithmetic throw when the currencies are different but only if both amounts are non-zero - e.g. it won't compare 2 EUR to 1 USD but a positive EUR is always greater than 0 USD, thus `anyMoney > Money.Zero` does not throw. 

> * `CountryCode` - if you use one of the ISO 3166-1 standards, countries have numeric codes but none have a code 0.

There is no need to have a country with code 0. For `Money`, I have something like this:

```c#
private readonly ICurrency _currency;
private readonly decimal _amount;
// ctors here...
public ICurrency Currency => _currency ?? Iso4217Currency.None;
```

Even though I validate currency against null in the constructors, `default(Money)` will always give me a null currency. So I default it to XXX on the property's getter and only access the currency through that property. Doing the same thing, you could have something like:

```c#
private readonly ushort _numericCode;
public int NumericCode => _numericCode != 0 ? _numericCode : DefaultNumericCode;
```

The problem here is that there is no _logical_ default value for a country; which country code do we want `default(CountryCode)` to represent? I think a class here would be more suitable.

I'm also not sure whether making it comparable makes sense as it is not clear to me whether it would compare on numeric or alphabetic codes. An alternative would be to expose static `IComparer<CountryCode>`s like `CountryCode.NumericComparer` and `CountryCode.AlphabeticComparer` if the type is going to be used in sorted sets/dictionaries a lot. And as long as you expose both codes publicly, clients can sort by the property they see fit anyway.

> * `Language` - ISO 639-* don't define numeric codes.

Again, I don't see this as a good type to be a struct as it does not have a logical default value. Like `CountryCode` (or say, my `Iso4217Currency`), `Language` is also a glorified enumeration that can be a class with cached instances.

```c#
var english = Language.GetLanguage("en");
```

> `Guid` (UUID) is a perfect example of that. Even though the UUID is defined as a 128-bit number and has a special "nil" case, it doesn't make sense to treat it as number.

Guid is a bit awkward as its default value is useless, but for everything else, it makes more sense to have as a struct than a class (small, immutable, equatable). So yes, despite my comments above, I do agree that there are valid use cases to have comparable, non-numeric structs.

The point I'm trying to make is that struct+IComparable is mainly a number trait. Yes, there are exceptions; but not so much. And even if there were more, the availability of these guards does not mean people will start using the "Positive" guard with their Color structs. I know it's not ideal, but I see this as the _slightly better_ compromise.

[1]: https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/structs#default-values
[2]: https://docs.microsoft.com/dotnet/standard/design-guidelines/struct

**Comment by girtsl** [comment]
Hi! 

**tl;dr**

It's always a good idea to do the following, especially for a general-purpose library:
- make the implicit explicit;
- remove or at least minimize ambiguity and guessing;
- evaluate which bits are general-purpose and which are specific to your application(s).

**The long version**

It's now clear to me that *for you*, in the context of structs, "Zero" is the same as the default value. I'm guessing it's because for a newly created struct instance all memory bits are set to zero. 

For me (and I suspect - for most people) "Zero" always means the number 0 and is distinct from a struct's default value for several reasons:
- different levels of abstraction: when writing reasonably high-level domain validation logic one does not usually think about such low level implementation details.
- "Zero" is singular whereas structs will most often consist of several members. This would imply that any struct can be folded to a single value - zero - which IMO doesn't make sense. One might argue that there are examples where the struct only contains a single numeric member (e.g. Money) but such examples are more often than not application specific and require some context.
- It doesn't always make sense to call the default value "Zero", a validator such as `Default` would be more generic and appropriate (in fact, you already have one with such name).
- If `Zero` means the same as `Default` - why have two methods (`Zero` and `Default`) that *seemingly* validate the same thing? I emphasized the word "seemingly" because in reality they don't - one uses the default comparer, the other - equality comparer.

On top of that, and as I've already pointed out, the documentation of methods `Zero`, `Positive`, etc. makes false claims:
- "Requires the argument value to be zero."
- "Requires the argument to have a value that is greater than zero."
- "Requires the argument to have a value that is less than zero."  

In reality those methods don't compare the argument to the number 0 but the result of invoking the default comparer.


Your example of `Money` struct and `Money.Zero` implementation is very application specific. For instance, your default `Money` instance `Money.Zero` has no currency which I don't consider a valid state. It's true that `any money > no money` but I'd treat any attempt to compare amounts in different currencies as an error unless an exchange rate lookup is involved. Again, very application-specific but generally makes the code a lot simpler and easier to reason about as there are fewer exceptions and assumptions. A general-purpose library should not include ambiguous and application-specific logic.

Perhaps my value type examples were contrived ones but they are not invalid just because there is no obvious *general-case* default value. In the scope of the application there might be perfectly valid default values.
The argument about a class vs a struct is not just about whether there is an obvious default value, it's more about whether you want it to be allocated on the heap or the stack, with a few other well documented differences.


I feel that this discussion has somewhat derailed from the original issue. Originally I made an argument for methods `GreaterThan` and `LessThan` in favor of `Min` and `Max` and then made the observation that `Positive` looks like a special case of `Min`. The observation turned out to be wrong but only because of the interpretation of what "Zero" means. The discussion then took a turn as I started to dig into those methods to see why my assumption was wrong.

To summarize: IMO your definition of "Zero" is application-specific and while it might make perfect sense for you, I think it doesn't apply to the generic-case, and is too ambiguous for a general-purpose library.

**P.S.**
A fun example to further illustrate the ambiguity of "zero", "positive", and "negative" and how it could trip unsuspecting developers:
```csharp
class Program
{
	static void Main(string[] args)
	{
		var _ = new Freezer(new Temperature {Degrees = -20, Scale = TemperatureScale.Celsius});
	}
}

public class Freezer
{
	public Temperature Temperature { get; }

	public Freezer(Temperature temperature)
	{
		Guard.Argument(() => temperature).Negative(); // a naive but natural assumption. also an ambiguous one
		
		Temperature = temperature;
	}
}

public struct Temperature : IComparable<Temperature>
{
	public decimal Degrees;
	public TemperatureScale Scale;
	public decimal Kelvins => ...; // degrees converted to Kelvin scale

	public int CompareTo(Temperature other)
	{
		return Kelvins.CompareTo(other.Kelvins);
	}
}

public enum TemperatureScale
{
	Kelvin = 0,
	Celsius,
	Fahrenheit,
	// ...
}
```

**Comment by safakgur** [comment]
Thank you for the detailed answer. I tried to design this with the struct best practices in mind, for the use cases that I thought would be the majority. But now it seems like the added value of those validations does not compensate the confusion they can create.

To sum up:

The following validations assume the default value of `IComparable<T>` structs to be _zero_: `Zero`/`NotZero`, `Negative`/`NotNegative`, `Positive`/`NotPositive`. And you'd prefer them to be either specific to `byte`, `sbyte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `float`, `double`, `TimeSpan`; or gone entirely, am I right?

`.GreaterThan(0)` is as readable as `.Positive()` if you ask me - so between the two options, we may as well delete the offenders altogether. So you'll do `.GreaterThan(new Money("GBP", 0M))` and I'll do `.GreaterThan(Money.Zero)` and everyone will be happy.

Temperature was an excellent example that I wouldn't have thought of myself. I would think a `Celsius` struct (or a `Temperature` that is documented to be always in Celsius) with methods like `FromKelvin`/`FromFahrenheit` (a bit like having an `RgbColor` and `CmykColor` instead of just `Color`, which would instead be a discriminated union of the other two in a language that supports it). Don't get me wrong, I'm definitely not saying this is the _one correct way_ - it's just not how I'd think of doing it, so it's great to see examples from other people.

**Comment by girtsl** [comment]
Apologies for the long delays between responses. I guess the global situation is taking a heavy toll on everyone right now, and will not end that soon.

```
Zero/NotZero, Negative/NotNegative, Positive/NotPositive. 
And you'd prefer them to be either specific or gone entirely, am I right?
```
Yes, I think that sums it up perfectly. As they are now, IMO, they are too ambiguous - I'd rather see them gone entirely and perhaps have the type-specific methods added later. An opportunity to further engage community, too.

```
So you'll do .GreaterThan(new Money("GBP", 0M)) and I'll do .GreaterThan(Money.Zero) 
and everyone will be happy.
```
That would make me very happy, thank you! I find that being explicit here aids clarity.

```
Temperature was an excellent example that I wouldn't have thought of myself.
```
TBH I'm playing the devil's advocate here and it makes me happy that I've been able to provide an example that has hit the bull's eye. It's not worth analyzing the example itself - it was meant to be short not a display of best practices - but I think it nicely exposes the ambiguity of what "zero" and the associated words "positive" and "negative" _can_ mean.
While to us the "obvious" default pick would be Celsius, to people elsewhere in the world it could be Fahrenheit, whereas to someone working in a scientific domain a more likely choice could be Kelvin - as that is the S.I. unit of temperature.

**Comment by pyrocumulus** [comment]
> It's only the `Min` and `Max` methods I would want to see replaced by `GreaterThanOrEqualTo` and `LessThanOrEqualTo` respectively.

I'd like to second this suggestion or at least try to look at providing the latter two as aliases for the former two. But I can understand having multiple methods doing the same work, is less nice in a way.

The api currently provides `GreaterThan`. When you are thinking in terms of if-statements and want to express a `GreaterThanOrEqualTo` type of validation on a `DateTime`, it feels awkward looking through the api not finding that validation. It took me quite a bit of time before understanding that `Min` was actually the same operation as the one I was looking for. It does not feel natural to me going from `GreaterThan` to `Min` while expressing almost the same validation on a date.

However, the `Min` and `Max` validations feel right in their own way. When I look at validations as a more mathematical operation, I can see those two feel more to the point. For example, all the validations I wrote on numeric values already used the `Min` validation already. That was before the above search happened. Funny how that works.

What I'm trying to say is; I can see both validations co-existing in Guard. But if you'd really want to support only one of those flavours, from a developers point-of-view, I'd suggest going for the `**OrEqualTo` types of validations and removing `Min/Max` counterparts. My two cents 🙂 

That being said, I love how expressive Guard is when doing validations for my own library. It makes argument validation less tedious and actually, slightly awesome 👍🏻

**Comment by safakgur** [comment]
Hi folks, thank you both for all your input, but unfortunately, I decided to retire Guard - I apologise for any inconvenience, and thank you for all the support. The readme has details.
